url433bec1e56_sarobe_DifficultyInvestigation_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/lunar/LunarProblem.java

```diff
@@ -0,0 +1,143 @@
+            // which is the reason for this constant
+            score = Math.abs(ship.pos.y - lt.getHeightAtX(ship.pos.x)) * 100;
+        }
+
+        return score;
+    }
+
+
+    static public double scoreTerminalState(Spaceship ship, LunarTerrain lt, double fuel) {
+        double score = 0;
+        // SCORE LANDING PAD PROXIMITY
+//        double dist1 = Math.abs(ship.pos.x - lt.getNearestSafeLandingPoint(ship.pos).x); // conventional distance
+//        double dist2 = Params.screenWidth - dist1; // wrapped distance
+//        double dist = Math.min(dist1, dist2); // closest of the two distances
+        double dist = Math.abs(ship.pos.x - lt.getNearestSafeLandingPoint(ship.pos).x);
+        // convert distance into a value between 0 and 1 by dividing by the maximum possible distance
+        double distScore = dist; // the highest possible horizontal distance
+        // SCORE VELOCITY
+        double vel = ship.vel.mag() - Params.survivableVelocity;
+        if(vel < 0) vel = 0;
+        double velScore = vel;
+        // SCORE FUEL
+        double fuelSpent = Params.startingFuel - fuel;
+        double fuelScore = fuelSpent;
+        // SCORE ANGLE
+        Vector2d shipHeading = new Vector2d(0, 1);
+        shipHeading.rotate(ship.rot);
+        double angDiff = shipHeading.angBetween(Params.landingFacing);
+        double angScore = angDiff;
+
+        double weighTotal = Params.proximityWeight + Params.velocityWeight + Params.fuelWeight + Params.angleWeight;
+        score = (distScore * Params.proximityWeight) + (velScore * Params.velocityWeight) + (fuelScore * Params.fuelWeight) + (angScore * Params.angleWeight);
+
+
+        return score;
+    }
+
+    public double[] getInstance(double[] x) {
+        return x;
+    }
+
+    public void demonstrationInit(List<Spaceship> ships, List<ShipController> conts, double pop[][]) {
+        shipFuels = new HashMap<Spaceship, Double>();
+        for(int i = 0; i < pop.length; i++) {
+            double[] s = pop[i];
+            Spaceship ship = new Spaceship();
+            ships.add(ship);
+            conts.add(new ShipController(ship, s));
+            ship.pos.set(Params.startingPoint);
+            shipFuels.put(ship, Params.startingFuel);
+        }
+    }
+
+    public void demonstrate(List<Spaceship> ships, List<ShipController> conts, double[][] pop) {
+        for(ShipController sc : conts) {
+            if(sc.getShip().isAlive())
+                sc.think();
+        }
+
+        synchronized (MainView.class) {
+            for(Spaceship s : ships) {
+                double fuel = shipFuels.get(s);
+                s.vel.add(0, Params.lunarGravity * Params.dt);
+                s.update();
+                s.pos.wrapHorizontal(Params.screenWidth);
+
+                // subtract fuel
+                fuel -= s.thrustForce;
+                shipFuels.put(s, fuel);
+
+                if(lt.isShipColliding(s) && s.label == "") {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.